### DISCUSSION TITLE
issue: defaultValues are set incorrectly if setValue is called immediately after reset

### DISCUSSION THREAD

**Discussion by andreev-artem** [react-hook-form/react-hook-form#10449] [Q&A]
### Version Number

7.43.9

### Codesandbox/Expo snack

https://codesandbox.io/s/condescending-boyd-k6cyt9?file=/src/App.tsx

### Steps to reproduce

1. Click on 'Submit'
2. In console `{test: "bar"}` as expected
3. Click on 'Clear'
4. Expected: empty input.
5. Click on 'Submit'
6. In console `{test: undefined}` as expected

### Expected behaviour

Expected behaviour - input is reset to the empty value (as specified in `reset`).

But for some reasons `defaultValue` for the input is set based on `control._formValues` here:
https://github.com/react-hook-form/react-hook-form/blob/v7.43.9/src/useController.ts#L60-L64

Shouldn't it be something like
```javascript
defaultValue: get(control._defaultValues, name, get(control._formValues, name, props.defaultValue))
```
instead?

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
I would not call setValue after reset, what's the use case? why not just provide value to reset.

**Comment by andreev-artem** [comment]
Use case: to be able to clear value in Dropdown with initial value. Actually for now there is inconsistence after Dropdown clear between value received on submit and value provided to Dropdown through `field.value`. I tried to illustrate that in [Codesandbox](https://codesandbox.io/s/condescending-boyd-k6cyt9?file=/src/App.tsx) and steps to reproduce.

But anyway is that correct to set default value for a field based on `control._formValues` instead of `control._defaultValues`? Change to that single line would lead to a more consistent behavior imho.

**Comment by bluebill1049** [comment]
> But anyway is that correct to set default value for a field based on control._formValues instead of control._defaultValues? Change to that single line would lead to a more consistent behavior IMHO.

It depends, user may want to preserve the input value even input gets unmounted at times.

**Comment by andreev-artem** [comment]
> It depends, user may want to preserve the input value even input gets unmounted at times.

So, user has logic on remove/add a control to DOM, but don't want to have logic to set initial value or somehow configure a control behavior on unmount? And that's why previously set `defaultValues` are ignored in favor of `control._formValues` while setting `defaultValue`. Sounds as minimum as implicit behavior.

Returning to the initial issue - why I can't `reset` defaultValues and immediately set initial value for a control? Why you don't consider that as a bug?

**Comment by bluebill1049** [comment]
> Returning to the initial issue - why I can't reset defaultValues and immediately set initial value for a control? Why you don't consider that as a bug?

send a PR and add some relevant test cases.

> So, user has logic on remove/add a control to DOM, but don't want to have logic to set initial value or somehow configure a control behavior on unmount? And that's why previously set defaultValues are ignored in favor of control._formValues while setting defaultValue. Sounds as minimum as implicit behavior.

If an input has value then it will render with its form value, it's as simple as that.

```tsx
reset({ whatever... })
setValue('data', 'value') // then it has form value now.
```

**Comment by andreev-artem** [comment]
> If an input has value then it will render with its form value, it's as simple as that.

Yeah, it's rendered as expected. But `defaultValue` set not as expected - `reset({ data: 'someDafaultValue' })` is just ignored.

**Comment by bluebill1049** [comment]
![Screenshot 2023-05-18 at 11 44 46 pm](https://github.com/react-hook-form/react-hook-form/assets/10513364/6ecce3d3-d120-486d-ad10-f18f1196c162)

it has the correct defaultValue what you have set in the `reset`

**Comment by andreev-artem** [comment]
> it has the correct defaultValue what you have set in the reset

Yeah, it sets it in one place, but ignores it later. I.e. it doesn't use it on Clear click. As result setting `defaultValues` don't work in that case.

I debugged a lot - internally in `useController` it ignores `defaultValues` because `control._formValues` has precedence over `control._defaultValues`. And changing mentioned line of code solved issue for me. I guess precedence order should be `props.defaultValue` => `control._defaultValues` => `control._formValues`. For now it's reverted.

Looks like issue title is a bit confusing, sorry for that, my bad.

**Comment by andreev-artem** [comment]
Works as expected on Clear: https://codesandbox.io/s/agitated-sun-dj901p?file=/src/App.tsx
Doesn't work on Clear: https://codesandbox.io/s/condescending-boyd-k6cyt9?file=/src/App.tsx

Difference is only - how `defaultValues` and initial value are set.

**Comment by bluebill1049** [comment]
`values` prop is calling `reset` at `useEffect`, so it's not a valid comparison. I am not sure what's the bug here, and what I have to fix.

**Comment by andreev-artem** [comment]
> values prop is calling reset at useEffect, so it's not a valid comparison. I am not sure what's the bug here, and what I have to fix.

Weird, for some reasons Clear button works correctly in that (first) case (control's value is set to default value, provided in `defaultValues`).  In the second case control's value is set to the value provided in `setValue` (not to the value provided in `reset`) on Clear button click.

I thought that's valid comparison because in both cases I provide `defaultValues` and set initial values - just using different approaches.

**Reply by bluebill1049** [reply]
values props will be reset at `useEffect`: https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L103-L109, while your example is reset and then setValue without `useEffect`

**Reply by andreev-artem** [reply]
Not sure I fully get you. Do you mean that developer should call API based on internal implementation (which may change without any warn) instead of docs? I.e. I understand difference from internals point of view, but internals may change without any warn but public interface shouldn't.

Based on docs and other issues `defaultValues` looks to me like values to which form inputs will be set on clear. And `useForm` `values` and `setValue` should set current input value and influence default values only if `defaultValues` wasn't provided. Isn't it? Based on that:

https://codesandbox.io/s/condescending-boyd-k6cyt9?file=/src/App.tsx - if it work not as expected per docs then it's bug. Isn't it?
